### ISSUE TITLE
PyPI Release: imaplib2

### ISSUE THREAD

**Issue by piyueh** [jazzband/help#226]
I'm working on the [jazzband guideline issue](https://github.com/jazzband/imaplib2/issues/13) for `imaplib2`. (I'm not a project leader, and there's no project leader for this project.) One of the tasks is to set up an auto-release mechanism using GitHub Actions. I have added a GitHub action workflow (see [here](https://github.com/jazzband/imaplib2/blob/master/.github/workflows/release.yml)), though I'm not sure if the workflow is correct or not. I would like to enable the auto-release. Thanks.

CC: @graingert

**Comment by jezdez** [comment]
Now that there is a project lead, this is ready to go from the infrastructure side. I would suggest to double-check locally that the source dist and wheel dist files created are what you are expecting. The package has a pretty simple layout nowdays, thanks for working on this!

**Comment by jezdez** [comment]
Oh, and @piyueh if you have questions about the actual steps of the release process, let me know and we can go over them together. Here's the primer:

1. Update version specifiers (maybe use [setuptools_scm](https://github.com/pypa/setuptools_scm) instead for that in the future?)
2. Push the changes to the repo and see if the tests pass.
3. Create a release through the GitHub release UI or create a tag and push it.
4. Once the release workflow has run its course, it'll upload the sdist and wheel dist files to the Jazzband package index.
5. Project leads will get an email asking to verify and confirm the release of the files to PyPI on the Jazzband website.

More info about the process can be found in the [Jazzband releases docs](https://jazzband.co/about/releases).

If you have any question, let me know.

**Comment by graingert** [comment]
we should double check the requires_python

**Comment by graingert** [comment]
yep we're good https://github.com/jazzband/imaplib2/blob/21ae2ed26d0ca66c683483d1bfdc7c3fb567a048/setup.cfg#L24

**Comment by graingert** [comment]
> Update version specifiers (maybe use setuptools_scm instead for that in the future?)

I don't like setuptools_scm because it stops github tar.gz archives from installing I prefer to use https://pypi.org/project/zest.releaser/ and would like to move to PEP 621 to set the version in the future

**Comment by jezdez** [comment]
> > Update version specifiers (maybe use setuptools_scm instead for that in the future?)
> 
> I don't like setuptools_scm because it stops github tar.gz archives from installing I prefer to use https://pypi.org/project/zest.releaser/ and would like to move to PEP 621 to set the version in the future

If installing directly from GitHub archive links is something you'd like to support, there is a way to do that using the [setuptools-scm-git-archive](https://pypi.org/project/setuptools-scm-git-archive/) project IIRC.

PEP 621 is certainly a worthwhile target, but it hasn't been adopted in the community. setuptools_scm on the other hand already has pyproject.toml support in case you want to make the jump already and may be a good step in between: https://github.com/pypa/setuptools_scm/#pyprojecttoml-usage I fully expect to evolve the packging config recommendation over the years for Jazzband projects, erring on the side of stability instead of novelty.

I can't say I have lots of experience with zest.releaser to be honest, but every time I've looked at it, it did too much for my taste. Granted, this is a personal preference. For Jazzband projects I've basically tried to recommend tools that are a bit more low-level on the UX scale and a "known evil" instead of an "unknown good".

**Comment by piyueh** [comment]
OK, these are what I'm going to do:

1. once https://github.com/jazzband/imaplib2/pull/27 is merged, I'll create a release for version 3.6 on GitHub (and hence a new release on PyPI). This will be the last version that does not have any changes in the source code.
2. close this issue if everything works fine.
3. convert the project to use `setuptools_scm` and remove all hard-coded `__version__` for future releases.

I've never used `zest.releaser`, but I agree it looks like an overkill. `imaplib2` currently is just a single-module package (and probably has only one application, `offlineimap`), so I guess we should refrain from overengineering it.

**Comment by graingert** [comment]
> If installing directly from GitHub archive links is something you'd like to support, there is a way to do that using the [setuptools-scm-git-archive](https://pypi.org/project/setuptools-scm-git-archive/) project IIRC.

That only works with tagged builds:

> Note that it only works for archives of tagged commits (because git currently lacks a format option equivalent to git describe --tags).

**Comment by piyueh** [comment]
The package has been successfully released on PyPI. The GitHub Action works fine. I'll close this issue to indicate the successful releasing.

The versioning issue will be tracked at https://github.com/jazzband/imaplib2/issues/29. Please feel free to re-open this issue if there're problems regarding the current releasing workflow.

@graingert I'm not familiar with the direct installation from a GitHub archive. If you think it's nice to have this feature, maybe we can open a new issue for it or discuss it at https://github.com/jazzband/imaplib2/issues/29?
